feat: Capability to retire tokens by serial number #4988

gayanath8 opened the pull request:
### Issue Label
[https://github.com/hashgraph/guardian/issues/4988](https://github.com/hashgraph/guardian/issues/4988)

### Problem Description
The current implementation of Hedera Guardian does not support retiring NFTs by a specific set of serial numbers, even though the Hedera SDK natively supports this operation. 

In applications such as carbon credit management, it is often necessary to retire selected tokens (e.g., specific serial numbers) rather than the entire supply or a fixed count.

### Requirements
Enhance Guardian to support NFT retirement by specifying a list of serial numbers. The implementation should:
* Allow users or policy workflows to pass a custom array of serial numbers.
* Call the appropriate Hedera SDK function to retire those specific serial-numbered NFTs.

### Summary of Changes
By applying this enhancement:

* Users can select specific NFT serial numbers to retire.
* For NFT operations:

  * **Serial Numbers** field is required.
  * Guardian will generate an array of serial numbers and pass them to the Hedera SDK for the wipe operation.
* For FT operations:

  * **Rule** remains the required field.
  * No changes are introduced to the FT wipe process.

### Validation Rules
* The Serial Numbers field is required.
* The Serial Numbers expression may contain individual numbers or inclusive ranges (e.g., 1,2-5,10).
* If a range is used (e.g., 2-5), the end value must be greater than or equal to the start value.
* Rule is required for FT operations.

### NFT Wipe Flow
**_Schema-Driven Wipe Flow_**

Serial numbers can be dynamically provided via schema fields or policy formulas. This enables multiple wipe operations for the same token, with different sets of serial numbers, as needed. Can useful for workflows requiring repeated or conditional NFT retirement.

<img width="886" height="3337" alt="FireShot Capture 004 - Guardian -  3 93 78 104" src="https://github.com/user-attachments/assets/e60da834-0507-461b-9dfa-a22a3e0b866f" />
<img width="787" height="287" alt="image" src="https://github.com/user-attachments/assets/ac6aaa2e-e8dd-4d5a-ab40-2787fbaed59a" />


* Example usage
  * Mint NFTs via the Mint Block, with the amount provided by the schema.
  * Add a separate schema for wipe operations.

<img width="985" height="300" alt="image" src="https://github.com/user-attachments/assets/022829b5-d344-4ec1-b26d-e299624cd899" />

  * Retire serial numbers 1, 3, 4, 5, 8, 12, 13, 14, 15, (entered dynamically in schema).

<img width="1098" height="832" alt="image" src="https://github.com/user-attachments/assets/d17401f6-3608-4d80-b3b4-3377bc8e4f40" />
<img width="1912" height="1907" alt="image" src="https://github.com/user-attachments/assets/9429d05c-ec0e-4bc3-abb8-6f2feb7b51af" />


**_Single Wipe Flow (Per User, Per Token)_**

A constant value for Serial Numbers Expression (for example, 1,2-5,10) can be provided. This allows a one-time wipe of specific NFTs per user and per token throughout the policy lifecycle.

<img width="788" height="283" alt="image" src="https://github.com/user-attachments/assets/60349543-c9e6-4a7f-ba74-5b1c1d2f07b9" />

* Example usage
  * Mint 20 NFTs using the Mint Block.
  * (Minting amount must be ≥ highest serial number to be retired)
  * Retire serial numbers 10,12,13,14,18. 
  
<img width="1790" height="828" alt="image" src="https://github.com/user-attachments/assets/da926dd8-19e8-4c52-a80d-4114887957de" />
<img width="1912" height="1457" alt="image" src="https://github.com/user-attachments/assets/b85f85e3-3e5d-481e-aeae-e4eeb94c1f7f" />


Transaction History of Single NFT
<img width="1847" height="420" alt="image" src="https://github.com/user-attachments/assets/793ad878-e0c2-4425-a1e2-6765b02dd7f1" />


### FT Wipe Flow

No changes to the existing FT token wipe flow.

<img width="975" height="410" alt="image" src="https://github.com/user-attachments/assets/f60f4776-5c50-4a80-baf0-474a842c8479" />

* Example usage
  * Create an FT token using schema-driven Mint Block.
  * Use Wipe Block to retire 1 token (based on Rule).
  
  
<img width="975" height="385" alt="image" src="https://github.com/user-attachments/assets/3bd6b26f-3656-4e37-a1d6-68387e944a14" />
<img width="975" height="425" alt="image" src="https://github.com/user-attachments/assets/4ef2fe95-7002-494d-95c0-e19592dfdb0f" />
<img width="1896" height="643" alt="image" src="https://github.com/user-attachments/assets/385cf245-3bf0-4866-8984-9a65801fffed" />

palindaa committed:
Update docker-compose-build.yml

Signed-off-by: palindaa <99331873+palindaa@users.noreply.github.com>

palindaa committed:
Merge pull request #1 from xeptagondev/palindaa-patch-1

Update docker-compose-build.yml

Neurone commented:
Hi, @gayanath8 thanks for the submission, really appreciated.

As mentioned in our last live meeting, it is important for the users to have more flexibility in the definition of the tokens to be wiped.

An idea is having a field where users can enter a range of serial numbers, such as "1,2-5,10" or "field0,field1-field5,field10," rather than just the starting and ending serial numbers.

What do you think? Can you update the PR accordingly?

The splitBlock probably already contains the code to handle both field values and fixed numbers (we can already do things like "field0/1000" there) so that can be a good starting point if you need info about how those things are managed.

Neurone commented:
As a side note, it seems the PR contains some changes unrelated to the specific features (for example the changes to the docker-compose-build.yml file). Can you please remove this kind of changes and focus only on the relevant ones?

It would be still useful gathering all the side improvements you think we should apply to other part of the source code, just use a separate PR for that.

anvabr reviewed (changes requested):
Can you please ensure that there are no errors displayed on the PR:
- need to fix DCO
- PR title to conform to 'conventional commits' spec
<img width="1467" height="319" alt="Screenshot 2025-10-09 at 16 06 03" src="https://github.com/user-attachments/assets/71b37528-3c72-4194-bfa2-15e95d2f0936" />

palindaa committed:
Capability to retire tokens by serial number Github Issue #4988

Signed-off-by: palindaa <palinda@xeptagon.com>

gayanath8 committed:
Added Wipe operation improvements
- Added UI policy validations for wipe block
- Added custom error notifications for NFT token wipes

Signed-off-by: gayanath8 <gayanathr@xeptagon.com>

palindaa committed:
Merge pull request #4 from xeptagondev/feature/4988-suggestion

Added Wipe operation improvements
